filebeat/reader/line/decoder.go

```diff
@@ -0,0 +1,166 @@
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package line
+
+import (
+	"fmt"
+	"io"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []uint8
+}
+
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) *decoderReader {
+	return &decoderReader{
+		in:         in,
+		decoder:    codec.NewDecoder(),
+		buf:        streambuf.New(nil),
+		encodedBuf: streambuf.New(nil),
+		bufferSize: bufferSize,
+	}
+}
+
+func (r *decoderReader) read(buf []byte) (int, error) {
+	b := make([]byte, r.bufferSize)
+
+	if r.buf.Len() != 0 {
+		return r.copyToOut(buf)
+	}
+
+	for {
+		start := r.encodedBuf.Len()
+		n, err := r.in.Read(b[start:])
+		if err != nil {
+			return len(buf[:n]), err
+		}
+
+		if start > 0 {
+			enc, err := r.encodedBuf.Collect(start)
+			if err != nil {
+				return 0, err
+			}
+			r.encodedBuf.Reset()
+			b = append(enc, b[start:]...)
+		}
+
+		nBytes, nProcessed, err := r.conv(b[:start+n], buf)
+		if err != nil {
+			if err == transform.ErrShortSrc {
+				r.encodedBuf.Append(b[nProcessed:])
+				return nBytes, nil
+			}
+			return 0, err
+		}
+		return nBytes, nil
+	}
+}
+
+// msgSize returns the size of the encoded message on the disk
+func (r *decoderReader) msgSize(symlen []uint8, size int) (int, []uint8, error) {
+	n := 0
+	for size > 0 {
+		if len(symlen) <= n {
+			return 0, symlen, fmt.Errorf("error calculating size: too short symlen")
+		}
+
+		size -= int(symlen[n])
+		n++
+	}
+
+	symlen = symlen[n:]
+
+	return n, symlen, nil
+}
+
+func (r *decoderReader) symbolsLen() []uint8 {
+	s := r.symlen
+	r.symlen = []uint8{}
+	return s
+}
+
+// conv converts encoded bytes into UTF-8 and produces a symlen array which
+// records the size of the encoded bytes and its converted size
+func (r *decoderReader) conv(in []byte, out []byte) (int, int, error) {
+	var err error
+	nProcessed := 0
+	decodedChar := make([]byte, 64)
+	r.symlen = make([]uint8, len(in))
+
+	i := 0
+	srcLen := len(in)
+	for i < srcLen {
+		j := i + 1
+
+		for j <= srcLen {
+			nDst, nSrc, err := r.decoder.Transform(decodedChar, in[i:j], false)
```

Comment:
ouch... checking `x/text`, there doesn't seem to be a `decodeRune` for any encoding. All encodings have the decode and copy functionality built right into the loop implemented by Transform :(
 
E.g. with `decodeRune` we could write the copy loop like this:

```
...
for i := 0; i < len(in); i++ {
  r, width, err := encoding.DecodeRune(in)
  if err != nil {
    ... // e.g. ErrShortSrc
  }

  uft8Width := utf8.RuneLen(r)
  if len(to) < utf8Width {
    ... // output buffer full, return
  }

  r.symLen[nProcessed] = width
  nProcessed++
  utf8.EncodeRune(to[nProcessed:], r)
}
...
```

We should extract the inner loop to decodeEncRune:

```
func decEncRune(decoder ..., in []byte) (rune, int, err) {
  for j := 0 ; j < len(in); j++ {
}
```

Reply:
Yes, this is the part of the code I am most worried about. It's complex and does not perform well. I would like to drop this algorithm in favour of the newer idea we discussed previously.